pkg/updatestrategy/drain.go

```diff
@@ -0,0 +1,450 @@
+
+func (m *KubernetesNodePoolManager) forceTerminationAllowed(pod v1.Pod, now, drainStart, lastForcedTermination time.Time) (bool, error) {
+	// too early to start force terminating
+	if now.Before(drainStart.Add(m.drainConfig.ForceEvictionGracePeriod)) {
+		m.podLogger(pod).Debugf("Won't force terminate (node in grace period)")
+		return false, nil
+	}
+
+	// we've recently force killed a pod
+	if now.Before(lastForcedTermination.Add(m.drainConfig.ForceEvictionInterval)) {
+		m.podLogger(pod).Debugf("Won't force terminate (recently force terminated a pod)")
+		return false, nil
+	}
+
+	// pod too young
+	if now.Before(pod.GetCreationTimestamp().Add(m.drainConfig.MinPodLifetime)) {
+		m.podLogger(pod).Debugf("Won't force terminate (pod too young)")
+		return false, nil
+	}
+
+	// find all other pods matched by the same PDBs
+	allPods, err := m.getPodsByNamespace(pod.GetNamespace())
+	if err != nil {
+		return false, err
+	}
+	allPdbs, err := m.getPDBsbyNamespace(pod.GetNamespace())
+	if err != nil {
+		return false, err
+	}
+	siblingPods := findSiblingPods(m.logger, pod, allPods.Items, allPdbs.Items)
+
+	// check if PDB siblings are old enough
+	siblingsOldEnough := true
+	for _, siblingPod := range siblingPods {
+		waitTime := m.drainConfig.MinUnhealthyPDBSiblingLifetime
+		if podReady(siblingPod) {
+			waitTime = m.drainConfig.MinHealthyPDBSiblingLifetime
+		}
+		if now.Before(siblingPod.GetCreationTimestamp().Add(waitTime)) {
+			siblingsOldEnough = false
+			break
+		}
+	}
+	if !siblingsOldEnough {
+		var siblingNames []string
+		for _, siblingPod := range siblingPods {
+			siblingNames = append(siblingNames, siblingPod.GetName())
+		}
+		m.podLogger(pod).Debugf("Won't force terminate (siblings %s too young)", strings.Join(siblingNames, ", "))
+		return false, nil
+	}
+
+	return true, nil
+}
+
+func podReady(pod v1.Pod) bool {
+	for _, condition := range pod.Status.Conditions {
+		if condition.Type == v1.PodReady {
+			return condition.Status == v1.ConditionTrue
+		}
+	}
+	return false
+}
+
+func findSiblingPods(logger *log.Entry, pod v1.Pod, allPods []v1.Pod, allPdbs []policy.PodDisruptionBudget) []v1.Pod {
+	var siblingSelectors []labels.Selector
+	for _, pdb := range allPdbs {
+		selector, err := metav1.LabelSelectorAsSelector(pdb.Spec.Selector)
+		if err != nil {
+			logger.Debugf("pdb %s/%s has an invalid selector: %v", pod.GetNamespace(), pdb.GetName(), err)
+			continue
+		}
+		if selector.Matches(labels.Set(pod.Labels)) {
+			siblingSelectors = append(siblingSelectors, selector)
+		}
+	}
+
+	var result []v1.Pod
+	for _, candidate := range allPods {
+		if candidate.GetName() == pod.GetName() {
+			continue
+		}
+
+		for _, selector := range siblingSelectors {
+			if selector.Matches(labels.Set(candidate.Labels)) {
+				result = append(result, candidate)
+				break
+			}
+		}
+	}
+
+	return result
+}
+
+func isPDBViolation(err error) bool {
+	return apiErrors.IsTooManyRequests(err) || strings.Contains(err.Error(), multiplePDBsErrMsg)
```

Comment:
Do I understand correctly that if apiserver returns 429 we will count this as PDB violation?
Why?

Reply:
This is the check we already had, so nothing changed here. We could also check the error message, but it's easier to just retry.